Tracking changes of stored entities(in session)

sashatim125 opened the pull request:
> there is a bug in session.save_changes() :
> custom metadata changes are not reflected in the command sent to the server
> 
fixed. Can be merged.

sashatim125 committed:
Tracking changes of stored entities(in session)

ayende commented on Raven.CppClient/BatchOperation.h:
```diff
@@ -11,7 +11,159 @@ namespace ravendb::client::documents::session::operations
 		size_t _session_commands_count{};
 		size_t _all_commands_count{};
-		std::unordered_map<std::string, std::shared_ptr<DocumentInfo>> _modifications{};
+		std::map<std::string, std::shared_ptr<DocumentInfo>, impl::utils::CompareStringsToIgnoreCase> _modifications{};
```
`CompareStringsToIgnoreCase` is an awkward name, use: `CompareStringsIgnoreCase`

sashatim125 replied:
OK, done

ayende commented on Raven.CppClient/InMemoryDocumentSessionOperations.h:
```diff
@@ -129,6 +136,87 @@ namespace ravendb::client::documents::session
+			{
+				return nullptr;
+			}
+			if (auto val_ptr = std::any_cast<std::string>(&object);
```
Wouldn't these just throw `bad_cast` ?

sashatim125 replied:
when casting from **a pointer** to `std::any` , a pointer is returned
with null instead of `std::bad_any_cast` in failure.
[https://en.cppreference.com/w/cpp/utility/any/any_cast](https://en.cppreference.com/w/cpp/utility/any/any_cast) (4)-(5)

sashatim125 replied:
I am considering using `std::variant` here.
[https://en.cppreference.com/w/cpp/utility/variant](https://en.cppreference.com/w/cpp/utility/variant)

sashatim125 committed:
EntityToJson added + small fixes

maximburyak commented on Raven.CppClient/BatchOperation.h:
```diff
@@ -11,7 +11,159 @@ namespace ravendb::client::documents::session::operations
+			if(!is_deferred)
+			{
+				entity = _entities.at(index);
+				if(const auto doc_info_it = session._documents_by_entity.find(entity);
```
Can we have this statement outside of if, just for clearer readability?

sashatim125 replied:
I'd like to keep it as is since that statement and the if condition are semantically coupled:
we ask if `_document_by_entity` contains the `entity` and if so, we update the `doc_info` variable.
The point is that STL containers do not have `contains()` method before c++20,
and that is the way to make it.

sashatim125 commented:
The bug mentioned above fixed.

maximburyak commented on Raven.CppClient/BatchOperation.h:
```diff
@@ -11,7 +11,159 @@ namespace ravendb::client::documents::session::operations
+				{
+					apply_metadata_modifications(id, modified_doc_info);
+				}
+			}else
```
allignment

sashatim125 replied:
OK

maximburyak commented on Raven.CppClient/BatchOperation.h:
```diff
@@ -11,7 +11,159 @@ namespace ravendb::client::documents::session::operations
+
+		void apply_metadata_modifications(const std::string& id, std::shared_ptr<DocumentInfo>doc_info)
+		{
+			doc_info->metadata_instance = std::static_pointer_cast<IMetadataDictionary>
```
I think that we lack something like      documentInfo.setMetadataInstance(metadata);
(https://github.com/ravendb/ravendb-jvm-client/blob/v4.1/src/main/java/net/ravendb/client/documents/session/operations/BatchOperation.java#L171)

sashatim125 replied:
We don't need it since we use a simple assignment.
All our auxiliary classes are defined as `struct`s with all members as public,
instead of private members and setters/getters.

maximburyak commented on Raven.CppClient/BatchOperation.h:
```diff
@@ -20,22 +172,139 @@ namespace ravendb::client::documents::session::operations
+			{
+				//TODO change to ClientVersionMismatchException
+				throw std::runtime_error("Cluster transaction was send to a node that is not supporting it. "
+					"So it was executed ONLY on the requested node on " /*+ _session.get().get_request_executor()->get_url()*/);
```
mark this one with TODO

sashatim125 replied:
OK

maximburyak commented on Raven.CppClient/BatchOperation.h:
```diff
@@ -20,22 +172,139 @@ namespace ravendb::client::documents::session::operations
+			if (auto&& res_size = result.results.size();
+				res_size >= _session_commands_count && res_size < _all_commands_count)
+			{
+				throw std::runtime_error("Batch command results are less then _all_commands_count");
```
@ayende should we define custom, typed exceptions for the cpp client?

sashatim125 replied:
@maximburyak do we have an agreement about that ?

sashatim125 committed:
DocumentStore added
BasicDocumentsSessionTests rewritten to use DocumentStore
various fixes

sashatim125 committed:
major refactoring
use DocumentSession instead of IDocumentSession
DocumentSession::advanced() creates AdvancedDocumentSessionOperations
AdvancedDocumentSessionOperations::extended() created ExtendedAdvancedSessionOperations
all of them use the same std::shared_ptr<DocumentSessionImpl>

maximburyak commented:
LGTM

sashatim125 committed:
"multyload" added

ayende commented on Raven.CppClient/AdvancedDocumentSessionOperations.h:
```diff
@@ -0,0 +1,164 @@
+	public:
+		~AdvancedDocumentSessionOperations() = default;
+
+		ExtendedAdvancedSessionOperations extended()
```
What is this, I'm not familiar with `ExtendedAdvancedSessionOperations`

sashatim125 replied:
there is `IAdvancedDocumentSessionOperations` and
`IAdvancedSessionOperations extends IAdvancedDocumentSessionOperations`.
so I thought it's a good idea to rename the `IAdvancedSessionOperations' to
'ExtendedAdvancedSessionOperations` which is way more meaningful.

ayende replied:
That introduce a new term into the API, and we want to have a single unified API nomenclature.

sashatim125 replied:
I see...
Since `IAdvancedDocumentSessionOperations` 's only usage is to be the super interface of
`IAdvancedSessionOperations`, could we just combine them into a single "interface"
`AdvancedSessionOperations` ?

maximburyak replied:
yeah, that'll be good. IAdvancedDocumentSessionOperations isn't really exposed as a property anywhere

maximburyak commented on Raven.CppClient.Tests/ds_definitions.h:
```diff
@@ -0,0 +1,70 @@
+			}else
+			{
+				_document_store.set_urls({ unsec_conn_details.get_url() });
+				_document_store.set_before_perform(set_for_fiddler);
```
does this mean we "always" use fiddler? in that case, we'll better have a flag that will control that

sashatim125 replied:
those are tests

sashatim125 replied:
The problem is we are using macros to ensure the uniqueness of the test's db name.
And because I don't want to multiply macros, I think we can use
`#define __USE_FIDDLER__` in every test suit that needs the fiddler.

maximburyak replied:
thing is that we don't want to even try to use fiddler, if we don't explicitly run in debug mode and if there is not fiddler process running, so we need to have some more checks here, and make sure that as long as we don't explicitly demand it, we don't even try using fiddler.

sashatim125 replied:
some refactoring was made so only the tests that `#define __USE_FIDDLER__`
at their top will use the Fiddler.

maximburyak commented on Raven.CppClient/BatchOperation.cpp:
```diff
@@ -0,0 +1,301 @@
+		session.increment_request_count();
+		_entities = result.entities;
+
+		return std::optional<commands::batches::BatchCommand>(std::in_place, /*_session.get().conventions*/conventions::DocumentConventions(),
```
Here we can use the store's conventions from the session

maximburyak commented on Raven.CppClient/BatchOperation.cpp:
```diff
@@ -0,0 +1,301 @@
+			const auto& batch_result = result.results[i];
+			if (!batch_result.is_object())
+			{
+				continue;//OK ?
```
looks legit

maximburyak commented on Raven.CppClient/BatchOperation.cpp:
```diff
@@ -0,0 +1,301 @@
+				handle_delete(batch_result);
+				break;
+			case commands::batches::CommandType::PATCH:
+				//TODO handle_patch(batch_result);
```
I think we better throw "not implemented" exceptions in addition to the todos, this way a "not working" test we'll encounter in the future will have a quick explanation on why it doesn't work

sashatim125 replied:
OK, but that would be during tryouts only, since when an exception is thrown during a test (run),
it just fails and we need to debug it anyway to find the source of the problem :(

maximburyak replied:
In the rest of the code you actully do throw "not implemented" exceptions, why did you decide avoiding it here?

sashatim125 replied:
@maximburyak you are right - the is no particular reason.
Fixed.

maximburyak commented on Raven.CppClient/BatchOperation.cpp:
```diff
@@ -0,0 +1,301 @@
+		//Do we need this??? Can we do this ???
+		if (entity)
+		{
+			//TODO _session.getGenerateEntityIdOnTheClient().trySetIdentity(entity, id);
```
We'll rely on the client's serialization/deserialization code to do that, unless you'll come to a conclusion it's not doable from there, and then we'll have to think about another solution
